src/ExpressionEvaluator/Core/Source/FunctionResolver/FunctionResolverBase.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Roslyn.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection.Metadata;
+
+namespace Microsoft.CodeAnalysis.ExpressionEvaluator
+{
+    internal abstract class FunctionResolverBase<TProcess, TModule, TRequest>
+        where TProcess : class
+        where TModule : class
+        where TRequest : FunctionResolutionRequestBase<TModule>
+    {
+        internal void EnableResolution(TProcess process, TRequest request)
+        {
+            var requests = GetRequests(process);
+            requests = requests.Push(request);
+            SetRequests(process, requests);
+            foreach (var module in GetAllModules(process))
+            {
+                ResolveRequests(requests, module);
+            }
+        }
+
+        internal void OnModuleLoad(TProcess process, TModule module)
+        {
+            var requests = GetRequests(process);
+            if (!requests.Any())
+            {
+                return;
+            }
+            ResolveRequests(requests, module);
+        }
+
+        internal abstract IEnumerable<TModule> GetAllModules(TProcess process);
+        internal abstract string GetModuleName(TModule module);
+        internal unsafe abstract bool GetModuleMetadata(TModule module, out byte* ptr, out int length);
```

Comment:
I think we could reduce unsafe code usage if this returned the MetadataReader directly, instead of ptr and length.